File: aoc-2020/rust/src/day1.rs
use std::fs;
use std::collections::HashSet;

pub fn part_one() -> i32 {
    let mut entries = HashSet::<i32>::new();
    let list = fs::read_to_string("../input/day1.txt").unwrap();
    let list = list.split("\n").map(|x| x.parse().unwrap());
    for a in list {
        let b = 2020 - a;
        if entries.contains(&b) {
            return a * b;
        } else {
            entries.insert(a);
        }
    }
    return -1;
}

pub fn part_two() -> i32{
    let mut entries = HashSet::<i32>::new();
    let list = fs::read_to_string("../input/day1.txt").unwrap();
    let list: Vec<i32> = list.split("\n").map(|x| x.parse().unwrap()).collect();
    for item in list.iter() {
        entries.insert(*item);
    };
    for a in list.iter() {
        for b in list.iter() {
            let c = 2020 - a - b;
            if entries.contains(&c) {
                return a * b * c;
            } 
        }
    }
    return -1;
}

File: aoc-2020/typescript/day1.ts
import fs from 'node:fs';

function partOne (lines: number[]) {
    for (const a of lines) {
        for (const b of lines) {
            if (a + b === 2020) {
                console.log(a*b)
                return         
            }
        }
    }
}

function partTwo(lines: number[]) {
    for (const a of lines) {
        for (const b of lines) {
            for (const c of lines) {
                if (a + b + c === 2020) {
                    console.log(a*b*c)
                    return         
                }
            }
        }
    }
}

fs.readFile(__dirname + '/../input/day1.txt', (_, data) => {
    const lines = data.toString()
        .split('\n')
        .map(Number)
    partOne(lines)
    partTwo(lines)
})
